agent/consul/acl.go

```diff
@@ -1592,6 +1592,13 @@ func vetDeregisterWithACL(rule acl.Authorizer, subj *structs.DeregisterRequest,
 	// We don't apply sentinel in this path, since at this time sentinel
 	// only applies to create and update operations.
 
+	// Allow service deregistration if the token has write permission for the node.
+	// This accounts for cases where the agent no longer has a token with write permission
+	// on the service to deregister it.
+	if rule.NodeWrite(subj.Node, nil) {
+		return nil
+	}
```

Comment:
Since we have this check now, further down there is another check for the inverse, which is no longer necessary: https://github.com/hashicorp/consul/blob/ff4150d5a7c451d847d63d403ce9242a9ae8a232/agent/consul/acl.go#L1626-L1628

I suggest instead just:

```golang
 return acl.ErrPermissionDenied 
 ```